crates/linker/src/macho.rs

```diff
@@ -601,453 +630,292 @@ fn gen_macho_le(
+    // Add a new text segment and data segment load commands.
+    // We assume existence of __ROC_TEXT,__text, __ROC_DATA_CONST,__const, __ROC_DATA,__data and __ROC_DATA,__bss.
     let segment_cmd_size = mem::size_of::<SegmentCommand64<LE>>();
     let section_size = mem::size_of::<Section64<LE>>();
+    let num_roc_cmds = 3u32;
+    let required_size = segment_cmd_size * num_roc_cmds as usize + section_size * 4;
 
     // We need the full command size, including the dynamic-length string at the end.
     // To get that, we need to load the command.
     let info = load_struct_inplace::<macho::LoadCommand<LE>>(exec_data, macho_load_so_offset);
     let total_cmd_size = info.cmdsize.get(LE) as usize;
 
-    // ======================== Important TODO ==========================
-    // TODO: we accidentally instroduced a big change here.
-    // We use a mix of added_bytes and md.added_byte_count.
-    // Theses should proabably be the same variable.
-    // Also, I just realized that I have been shifting a lot of virtual offsets.
-    // This should not be necessary. If we add a fully 4k of buffer to the file, all of the virtual offsets will still stay aligned.
-    // So a lot of the following work can probably be commented out if we fix that.
-    // Of coruse, it will cost about 4k bytes to a final executable.
-    // This is what the elf version currently does.
-    // Theoretically it is not needed (if we update all virtual addresses in the binary),
-    // but I definitely ran into problems with elf when not adding this extra buffering.
-
-    // Copy header and shift everything to enable more program sections.
-    let added_byte_count = ((2 * segment_cmd_size) + (2 * section_size) - total_cmd_size) as u64;
-    md.added_byte_count = added_byte_count
-        // add some alignment padding
-        + (MIN_SECTION_ALIGNMENT as u64 - added_byte_count % MIN_SECTION_ALIGNMENT as u64);
-
-    md.exec_len = exec_data.len() as u64 + md.added_byte_count;
+    let available_size =
+        md.start_of_first_section - (size_of_cmds + size_of_header - total_cmd_size);
 
-    let mut out_mmap = open_mmap_mut(out_filename, md.exec_len as usize);
-    let end_of_cmds = size_of_cmds + mem::size_of_val(exec_header);
-
-    // "Delete" the dylib load command - by copying all the bytes before it
-    // and all the bytes after it, while skipping over its bytes.
-    // It has a dynamic-length string at the end that we also need to delete,
-    // in addition to the header.
-    out_mmap[..macho_load_so_offset].copy_from_slice(&exec_data[..macho_load_so_offset]);
-
-    out_mmap[macho_load_so_offset..end_of_cmds - total_cmd_size]
-        .copy_from_slice(&exec_data[macho_load_so_offset + total_cmd_size..end_of_cmds]);
+    if available_size < required_size {
+        internal_error!("Not enough free space between end of load commands and start of first section\nConsider recompiling the host with -headerpad <size> linker flag");
```

Comment:
I need to double check if elf linkers have a equivalent flag. Would simplify a lot of the surgical linker wrangling if so.

Reply:
I don't think there was anything like this but you should check yeah. So if I understand it right you would like to reserve some space between the end of the program headers table and the beginning of the first section, correct? You would also need to take care of extending the read-only segment by some amount if the table is enclosed in that segment. I remember doing a trick where Zig's linker would put the program header table in its own read-only segment, perhaps that could somehow be accomplished, if not with flags, then with the dreaded linker scripts? Anyhow, happy brainstorm more if you want!